File: ConicBundle/cppinterface/cb_psqmr.cpp
dll void cb_psqmr_destroy(Psqmr* self) {
  delete self;
}

dll Psqmr* cb_psqmr_new(int pril = -1) {
  return new Psqmr(&std::cout, pril);
}

dll void cb_psqmr_set_maxit(Psqmr* self, Integer in_maxit) {
  self->set_maxit(in_maxit);
}

dll Integer cb_psqmr_get_maxit(const Psqmr* self) {
  return self->get_maxit();
}

dll int cb_psqmr_get_err(const Psqmr* self) {
  return self->get_err();
}

dll Integer cb_psqmr_get_nmult(const Psqmr* self) {
  return self->get_nmult();
}

dll Real cb_psqmr_get_residual_norm(const Psqmr* self) {
  return self->get_residual_norm();
}

dll Real cb_psqmr_get_avg_reduction(const Psqmr* self) {
  return self->get_avg_reduction();
}

dll Real cb_psqmr_get_termprec(const Psqmr* self) {
  return self->get_termprec();
}

dll int cb_psqmr_compute(Psqmr* self, IterativeSystemObject* system, Matrix* x, Real termprec, Matrix* storex = 0, Integer storestep = 0) {
  return self->compute(*system, *x, termprec, storex, storestep);
}

dll void cb_psqmr_set_out(Psqmr* self, int in_print_level = 1) {
  self->set_out(&std::cout, in_print_level);
}

File: src/cppinterface/cb_psqmr.jl
@doc raw"""
    CBPsqmr(pril::Integer = -1)

default constructor
"""
CBPsqmr(pril::Integer = -1) = CBPsqmr(@ccall libcb.cb_psqmr_new(pril::Cint)::Ptr{Cvoid})

@doc raw"""
    cb_set_maxit!(self::CBPsqmr, in_maxit::Integer)

set maximum number of iterations
"""
cb_set_maxit!(self::CBPsqmr, in_maxit::Integer) = @ccall libcb.cb_psqmr_set_maxit(self.data::Ptr{Cvoid}, in_maxit::Cint)::Cvoid

@doc raw"""
    cb_get_maxit(self::CBPsqmr)

get maximum number of iterations
"""
cb_get_maxit(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_maxit(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_err(self::CBPsqmr)

returns the error code of the last call
"""
cb_get_err(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_err(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_nmult(self::CBPsqmr)

returns the number of matrix-vector multiplications of the last call
"""
cb_get_nmult(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_nmult(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_residual_norm(self::CBPsqmr)

returns the residual norm of last call
"""
cb_get_residual_norm(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_residual_norm(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_avg_reduction(self::CBPsqmr)

returns the average of the achieved reduction factor per iteration
"""
cb_get_avg_reduction(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_avg_reduction(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_termprec(self::CBPsqmr)

return the (absolute) precision requirement for termination used in the last call
"""
cb_get_termprec(self::CBPsqmr) = @ccall libcb.cb_psqmr_get_termprec(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_compute!(self::CBPsqmr, system::CBIterativeSystemObject, x::CBMatrix, termprec::Real, storex::Union{<:CBMatrix,Nothing} = nothing, storestep::Integer = 0)

compute the solution for system into x with (absolute) residual precision termprec
"""
cb_compute!(self::CBPsqmr, system::CBIterativeSystemObject, x::CBMatrix, termprec::Real, storex::Union{<:CBMatrix,Nothing} = nothing, storestep::Integer = 0) = @ccall libcb.cb_psqmr_compute(self.data::Ptr{Cvoid}, system.data::Ptr{Cvoid}, x.data::Ptr{Cvoid}, termprec::Cdouble, (isnothing(storex) ? C_NULL : storex.data)::Ptr{Cvoid}, storestep::Cint)::Cint

@doc raw"""
    cb_set_out!(self::CBPsqmr, in_print_level::Integer = 1)

set output stream and level of detail of log output (for debugging)
"""
cb_set_out!(self::CBPsqmr, in_print_level::Integer = 1) = @ccall libcb.cb_psqmr_set_out(self.data::Ptr{Cvoid}, in_print_level::Cint)::Cvoid

